workflow.go

```diff
@@ -0,0 +1,161 @@
+
+				if cancel.CancelSubscription {
+					// TODO: cancel subscription
+				}
+			}).
+			Select(ctx)
+	}
+}
+
+func pollHistorySize(ctx workflow.Context) {
+	logger := workflow.GetLogger(ctx)
+	info := workflow.GetInfo(ctx)
+
+	for {
+		_ = workflow.Sleep(ctx, time.Duration(HISTORY_CHECK_PERIOD_SECONDS*int(time.Second)))
+
+		logger.Info("Current history length", "History Length", info.GetCurrentHistoryLength())
+		if info.GetCurrentHistoryLength() > EVENT_HISTORY_THRESHOLD {
+			CONTINUE_AS_NEW_CHANNEL.Send(ctx, true)
+		}
+	}
+}
+
+func SubscriptionWorkflow(ctx workflow.Context, subscription SubscriptionInfo) error {
+	ao := workflow.ActivityOptions{
+		StartToCloseTimeout: 10 * time.Second,
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+
+	logger := workflow.GetLogger(ctx)
+	logger.Info("Subscription workflow started", "customer", subscription.Customer, "subscription", subscription)
+
+	// Immediately kick off update/cancel handlers, since those can come in at
+	// any time. Channel used to signal that Continue-As-New is needed
+	var m Messages
+	CONTINUE_AS_NEW_CHANNEL = workflow.NewChannel(ctx)
+	workflow.GoNamed(ctx, "signalPoller", func(ctx workflow.Context) {
+		m.HandleSignals(ctx, CONTINUE_AS_NEW_CHANNEL)
+	})
+
+	continueAsNew := false
+	continueHandler := func(c workflow.ReceiveChannel, _ bool) {
+		var canMessage bool
+		c.Receive(ctx, &canMessage)
+		if canMessage {
+			logger.Info("History size too big. time to CAN")
+			// TODO continue-as-new
+			continueAsNew = true
+		}
+	}
+	workflow.GoNamed(ctx, "historyPoller", pollHistorySize)
+
+	var a *Activities
+	err := workflow.ExecuteActivity(ctx, a.SendWelcomeEmail, subscription).Get(ctx, nil)
+	if err != nil {
+		logger.Error("Welcome Email Activity failed.", "Error", err)
+		return err
+	}
+
+	// Start trial period timer, async
+	trialDuration := time.Duration(subscription.TrialPeriodDays * 24 * float64(time.Hour))
+	trialTimer := workflow.NewTimer(ctx, trialDuration)
+	expectedTimerFire := workflow.Now(ctx).Add(trialDuration)
+
+	workflow.NewSelector(ctx).
+		AddFuture(trialTimer, func(f workflow.Future) {
+			logger.Info("Trial has ended. Sending 'trial is over' email, then entering normal billing cycle.")
+		}).
+		AddReceive(CONTINUE_AS_NEW_CHANNEL, continueHandler).
+		// wait for update/cancel signals, Continue-As-New channel, or Trial timer.
+		Select(ctx)
+
+	// if we're here because of Continue-As-New, then do that. Otherwise continue as normal.
+	if continueAsNew {
+		remainingTime := expectedTimerFire.Sub(workflow.Now(ctx))
+		logger.Info("Continuing as New.", "Remaining time in trial period", remainingTime)
+
+		subscription.TrialPeriodRemaining = remainingTime
+		canError := workflow.NewContinueAsNewError(ctx, SubscriptionWorkflow, subscription)
+		return canError
+	}
+
+	subscription.TrialPeriodRemaining = time.Duration(0)
+	workflow.ExecuteActivity(ctx, a.SendTrialExpiredEmail, subscription).Get(ctx, nil)
```

Comment:
Mentioned before (will stop mentioning), but it is really important to handle all errors